store/model.go

```diff
@@ -0,0 +1,276 @@
+		return txn.Save(v)
+	})
+}
+
+func (m *Model) Has(id es.EntityID) (exists bool, err error) {
+	m.ReadTxn(func(txn *Txn) error {
+		exists, err = txn.Has(id)
+		return err
+	})
+	return
+}
+
+type Txn struct {
+	model     *Model
+	discarded bool
+	commited  bool
+	readonly  bool
+
+	actions []es.Action
+}
+
+type SaveOp struct {
+	Before interface{}
+	After  interface{}
+}
+
+func (t *Txn) Create(new interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	id := getEntityID(new)
+	if id == es.EmptyEntityID {
+		id = setNewEntityID(new)
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return errCantCreateExistingInstance
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Save(updated interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+
+	id := getEntityID(updated)
+	key := t.model.dsKey.ChildString(id.String())
+	beforeBytes, err := t.model.datastore.Get(key)
+	if err == ds.ErrNotFound {
+		return errCantSaveNonExistentInstance
+	}
+	if err != nil {
+		return err
+	}
+
+	before := reflect.New(t.model.valueType.Elem()).Interface()
+	err = json.Unmarshal(beforeBytes, before)
+	if err != nil {
+		return err
+	}
+	a := es.Action{
+		Type:       es.Save,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   before,
+		Current:    updated,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Delete(id es.EntityID) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return ErrNotFound
+	}
+	a := es.Action{
+		Type:       es.Delete,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    nil,
+	}
+	t.actions = append(t.actions, a)
+	return nil
+}
+
+func (t *Txn) Has(id es.EntityID) (bool, error) {
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return false, err
+	}
+	return exists, nil
+}
+
+func (t *Txn) FindByID(id es.EntityID, v interface{}) error {
+	key := t.model.dsKey.ChildString(id.String())
+	bytes, err := t.model.datastore.Get(key)
+	if errors.Is(err, ds.ErrNotFound) {
+		return ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(bytes, v)
+}
+
+func (t *Txn) Commit() error {
+	if t.discarded || t.commited {
+		return errAlreadyDiscardedCommitedTxn
+	}
+
+	events, err := t.model.eventcreator.Create(t.actions)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		if err := t.model.dispatcher.Dispatch(e); err != nil {
+			return err // ToDo/Note: Important to document the implications of a partial dispatch
```

Comment:
Important comment here... I'd like to think about what should happen here. Not really trivial.

Reply:
Yeh, ideally, you'd rollback any transactions that resulted from the dispatch... this is something I really struggled with previously, and never had a good solution. I think I convinced myself that partial dispatches were ok... but not so sure about that? Also, I think we could leave some of these decisions up to the caller?

Reply:
Created #15 to keep track of this topic since is quite complex. I tried to write there a summary of the problem and some ideas that are now in my mind, feel free to add more points there if something extra is fresh now in yours minds.